app/controllers/provider_interface/provider_agreements_controller.rb

```diff
@@ -0,0 +1,55 @@
+module ProviderInterface
+  class ProviderAgreementsController < ProviderInterfaceController
+    before_action :set_page_name
+    skip_before_action :check_data_sharing_agreements
+
+    def new_data_sharing_agreement
+      @provider_agreement = GetPendingDataSharingAgreementsForProviderUser.call(provider_user: current_provider_user).first
+      if @provider_agreement
+        render_agreement
+      else
+        redirect_to action: :show_data_sharing_agreement
+      end
+    end
+
+    def create_data_sharing_agreement
+      @provider_agreement = ProviderAgreement.new(provider_agreement_params.merge(provider_user: current_provider_user))
+      begin
+        @provider_agreement.save!
+        redirect_to '/provider'
+      rescue ActiveRecord::RecordInvalid
+        render_agreement
+      end
+    end
+
+  private
+
+    def set_page_name
+      @page_name = 'data_sharing_agreement'
+    end
+
+    def set_declaration
+      @declaration = \
+        if !@provider_agreement.persisted?
+          "#{@organisation_name} agrees to comply with the data sharing practices outlined in this agreement"
+        else
+          agreement_date = @provider_agreement.accepted_at.strftime("%d %B %Y")
+          "#{@organisation_name} agreed to comply with the data sharing practices outlined in this agreement on #{agreement_date}"
+        end
+    end
+
+    def render_agreement
+      @organisation_name = @provider_agreement.provider.name
+      set_declaration
```

Comment:
Perhaps it would be cleaner to use the `@provider_agreement` var in the views, instead of generating new instance variables here.

Reply:
yes